src/sio/sio.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: sio.3,v 1.2 2003/06/17 05:10:54 seth Exp $
.TH SIO 3X "29 May 1992"
.SH NAME
Sread, Sgetc, Srdline, Sfetch, Swrite, Sputc, Sprint, Sprintv, Sdone, Sundo, Stie, Suntie, Sflush, Sclose, Sbuftype, Smorefds, Sgetchar, Sputchar, SIOLINELEN, SIOMAXLINELEN - fast stream I/O
.SH SYNOPSIS
.LP
.nf
.ft B
#include "sio.h"
.LP
.ft B
int Sread( fd, buf, nbytes )
int fd ;
char *buf ;
int nbytes ;
.LP
.ft B
int Sgetc( fd )
int fd ;
.LP
.ft B
char *Srdline( fd )
int fd ;
.LP
.ft B
char *Sfetch( fd, length )
int fd ;
long *length ;
.LP
.ft B
int Swrite( fd, buf, nbytes )
int fd ;
char *buf ;
int nbytes ;
.LP
.ft B
int Sputc( fd, c )
int fd ;
char c ;
.LP
.ft B
int Sprint( fd, format [ , ... ] )
int fd ;
char *format ;
.LP
.ft B
int Sprintv( fd, format, ap )
int fd ;
char *format ;
va_list ap ;
.LP
.ft B
int Sdone( fd )
int fd ;
.LP
.ft B
int Sundo( fd, type )
int fd ;
int type ;
.LP
.ft B
int Stie( ifd, ofd )
int ifd, ofd ;
.LP
.ft B
int Suntie( fd )
int fd ;
.LP
.ft B
int Sbuftype( fd, type )
int fd, type ;
.LP
.ft B
int Smorefds()
.LP
.ft B
int Sflush( fd )
int fd ;
.LP
.ft B
int Sclose( fd )
int fd ;
.LP
.ft B
int Sgetchar( fd )
int fd ;
.LP
.ft B
int Sputchar( fd, c )
int fd;
char c ;
.LP
.ft B
int SIOLINELEN( fd )
int fd ;
.LP
.ft B
int SIOMAXLINELEN( fd )
int fd ;
.SH DESCRIPTION
The \fISIO\fR library provides support
for \fIstream\fR I/O on file descriptors.
The first argument of every function
or macro is a file descriptor. The file descriptor may be used either for
input or for output, but not both. Attempting to use a descriptor for
both input and output will cause the call for the latter use to fail.
When you are
done with using a file descriptor, you should inform \fISIO\fR
by invoking \fBSdone()\fR (unless the program is about to
call \fIexit(3)\fR).
You can also use \fBSdone()\fR if
you want to perform a different type of operation on the same
file descriptor (e.g. first you were reading data from the file
descriptor and then you want to write some data).
Another possibility is to do stream I/O at different file offsets
by using \fBSdone()\fR before using \fBlseek(2)\fR to move to a
new file offset.
.LP
I/O operations on different file descriptors do not interfere
(unless the file descriptors refer to the same file, in which case
the results are undefined).
.LP
For disk files, I/O always starts at the current file offset.
If that offset is not a multiple of the preferred block size for file
system I/O, performance will not be optimal
(the preferred block size is determined from the
\fIst_blksize\fR field in \fIstruct stat\fR).
For optimal performance, it is recommended that no I/O operations
(like \fIread(2)\fR or \fIwrite(2)\fR)
are applied to the file descriptor if it is to be used by \fISIO\fR.
.LP
Read I/O is either buffered, or is done using memory mapping whenever
that is possible and appropriate.
.LP
The library functions that do stream I/O resemble system calls
(for example \fBSread()\fR resembles \fIread(2)\fR) so that modifying
a program that uses the system calls to use the \fISIO\fR functions
is easy (e.g. just replace \fIread(2)\fR with \fBSread()\fR; the function
signatures as well as the return values are exactly the same; also make
sure to replace calls to \fIclose(2)\fP with \fBSclose()\fP).
.LP
\fISIO\fR uses the underlying system calls \fIread(2)\fR and \fIwrite(2)\fR
to do I/O (except when reading files using memory mapping).
These calls may be interrupted (i.e. they may return -1 with
.I errno
set to EINTR). Such interruptions are ignored by \fISIO\fR which
simply reissues the system call
(this means that a \fISIO\fP call will never fail because the
underlying I/O system call was interrupted).
.LP
.B Sread()
reads \fInbytes\fR bytes from the stream associated with file
descriptor \fIfd\fR into the buffer pointed to by \fIbuf\fR.
.LP
.B Sgetc()
reads a character from the stream
associated with file descriptor \fIfd\fR.
It returns \fBSIO_EOF\fR if the end of file has been reached.
.LP
.B Sgetchar()
(a macro) performs exactly the same function as \fBSgetc()\fR but
it is much faster.
.LP
.B Srdline()
reads a line from the stream
associated with file descriptor \fIfd\fR.
The newline at the end of the line is replaced by a NUL byte. Lines
longer than the maximum line length supported by \fISIO\fR will
have characters deleted.
.LP
.B SIOLINELEN()
(a macro) returns the length of
the line returned by the last call to \fBSrdline()\fR
(the value returned by \fBSIOLINELEN()\fR is valid only after
\fBSrdline()\fR and as long as no other
\fISIO\fR calls are performed on that file descriptor).
.LP
.B SIOMAXLINELEN()
(a macro) returns
the maximul line length supported by \fISIO\fR for the file
descriptor. As a side-effect it initializes \fIfd\fR for input.
.LP
.B Sfetch()
returns a pointer to data coming from the stream
associated with file
descriptor \fIfd\fR. The amount of data available is indicated
by the \fIlength\fR argument. One possible use for this function
is to copy files.
.LP
.B Swrite()
writes \fInbytes\fR bytes to the stream associated with file
descriptor \fIfd\fR from the buffer pointed to by \fIbuf\fR.
.LP
.B Sputc()
writes a single character to the stream
associated with file descriptor \fIfd\fR.
.LP
.B Sputchar()
(a macro) performs exactly the same function as \fBSputc()\fR
but it is much faster.
.LP
.B Sprint()
imitates the behavior of printf(3) as defined in the
ANSI C Standard. There are some limitations. Check the \fBSprint()\fR
man page for more information.
.LP
.B Sprintv()
is the same as \fBSprint()\fR except that it uses a
\fIvarargs\fR argument list.
.LP
.B Sundo()
returns the characters returned by the last call to
\fBSrdline()\fR, \fBSgetc()\fR or \fBSgetchar()\fR to the stream
so that they can be reread. The \fItype\fR argument to \fBSundo()\fR
can be \fBSIO_UNDO_LINE\fR or \fBSIO_UNDO_CHAR\fR depending
on whether the call whose effect needs to be undone was
\fBSrdline()\fR or \fBSgetc()\fR/\fBSgetchar()\fR respectively.
There is no check on
whether the last function invoked on \fIfd\fR was one of the above
and the results are undefined if there is no correspondence
between the \fItype\fR and the last operation on \fIfd\fR.
(i.e. the result is undefined if you try \fBSIO_UNDO_CHAR\fR
and the last operation was not \fBSgetchar()\fR or \fBSgetc()\fR).
.LP
.B Stie()
ties the file descriptor \fIifd\fR to the file descriptor \fIofd\fR.
This means that whenever a \fIread(2)\fR is done on \fIifd\fR, it is
preceded by a \fIwrite(2)\fR on \fIofd\fR.
For filters it is useful to do \fIStie( 0, 1 )\fR to maximize concurrency.
It is also useful to do the same thing when you issue prompts to the
user and you want the user reply to appear on the same line with the
prompt.
\fIifd\fR, \fIofd\fR  will be initialized for input, output respectively
(if any of them is initialized, it must be for the appropriate
stream type (input or output)).
If \fIifd\fR was tied to another file descriptor, the old tie is broken.
.LP
.B Suntie()
undoes the effect of \fBStie()\fR for the specified input file descriptor.
.LP
.B Sbuftype()
determines the buffering type for the output stream associated with
file descriptor \fIfd\fR.
By default output directed to terminals is line buffered, output
directed to file descriptor 2 (standard error) is unbuffered and
everything else is fully buffered.
Possible values for the \fItype\fR argument are
.RS
.TP 15
.SB SIO_FULLBUF
for full buffering
.TP
.SB SIO_LINEBUF
for line buffering
.TP
.SB SIO_NOBUF
for no buffering
.RE
.LP
.B Smorefds()
should be used to inform \fBSIO\fR that the number of available file
descriptors has been increased. \fBSIO\fR uses an array of internal
stream descriptors which are indexed by the file descriptor number. Some
operating systems (ex. SunOS 4.1[.x]) allow the number of available
file descriptors to vary. If that number is increased beyond its initial
value \fBSIO\fR needs to know in order to allocate more stream descriptors.
.LP
.B Sdone()
flushes any buffered output for \fIfd\fR
and releases the \fISIO\fR resources used. \fBSdone()\fR
is useful in case the program needs to reprocess the
data of a file descriptor (assuming the file descriptor corresponds
to a file).  The program can call \fBSdone()\fR,
\fIlseek(2)\fR to the beginning of the file
and then proceed to reread the file.
.LP
.B Sflush()
causes any buffered stream output to be written to the
file descriptor. If its argument is the special value \fBSIO_FLUSH_ALL\fR
then all output streams will be flushed.
.LP
.B Sclose()
closes a file descriptor used for stream I/O, flushes
any buffered output and releases the \fISIO\fR resources used.
.SH EXAMPLES
.LP
The following code implements a (poor) substitute for the tee command
(it copies standard input to a file as well as to standard output).
.ne 10
.RS
.nf
.ft B
#include "sio.h"
.sp .5
main( argc, argv )
	int argc ;
	char *argv[] ;
{
	char *file = (argc > 1) ? argv[ 1 ] : "tee.file" ;
	int fd = creat( file, 0644 ) ;
	long length ;
	char *s ;
.sp .5
	while ( s = Sfetch( 0, &length ) )
	{
		Swrite( 1, s, length ) ;
		Swrite( fd, s, length ) ;
	}
	exit( 0 ) ;
}
.fi
.ft R
.RE
.SH RETURN VALUES
.LP
.B Sread()
returns the number of bytes read on success
(0 means end-of-file)
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sgetc()
returns the character read on success,
SIO_EOF when the end-of-file is reached,
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Srdline()

ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive

src/sio/sio.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */

static const char RCSid[] = "$Id: sio.c,v 1.3 2003/06/17 05:10:54 seth Exp $";
static const char sio_version[] = VERSION;

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>

#if defined (_WIN32) && !defined(__CYGWIN__)
# include <sosnt.h>
#endif /* _WIN32 && !__CYGWIN__ */

#include "sio.h"
#include "impl.h"

#ifdef EVENTS
#include "events.h"
#endif
#include "clchack.h"


/*
 * SIO WRITE FUNCTIONS: Swrite, Sputc
 */

/*
 * Stream write call: arguments same as those of write(2)
 */
int Swrite(int fd, register char *addr, register int nbytes)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_od_t *odp = ODP( dp ) ;
	register int b_transferred ;
	register int b_avail ;
	int total_b_transferred ;
	int b_written ;
	int b_in_buffer ;

#ifdef EVENTS
	EVENT( fd, EV_SWRITE ) ;
#endif

	IO_SETUP( fd, dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;
	ASSERT( odp->start <= odp->nextb && odp->nextb <= odp->buf_end ) ;

	b_avail = odp->buf_end - odp->nextb ;
	b_transferred = MIN( nbytes, b_avail ) ;
	sio_memcopy( addr, odp->nextb, b_transferred ) ;
	odp->nextb += b_transferred ;

	/*
	 * check if we are done
	 */
	if ( b_transferred == nbytes )
		return( b_transferred ) ;

	/*
	 * at this point we know that the buffer is full
	 */
	b_in_buffer = odp->buf_end - odp->start ;
	b_written = __sio_writef( odp, fd ) ;
	if ( b_written != b_in_buffer )
		return( (b_written >= nbytes) ? nbytes : b_written ) ;

	total_b_transferred = b_transferred ;
	addr += b_transferred ;
	nbytes -= b_transferred ;

	for ( ;; )
	{
		b_transferred = MIN( nbytes, odp->buffer_size ) ;
		sio_memcopy( addr, odp->nextb, b_transferred ) ;
		odp->nextb += b_transferred ;
		nbytes -= b_transferred ;
		if ( nbytes == 0 )
		{
			total_b_transferred += b_transferred ;
			break ;
		}
		/*
		 * the buffer is full
		 */
		b_written = __sio_writef( odp, fd ) ;
		if ( b_written != odp->buffer_size )
		{
			if ( b_written != SIO_ERR )
			{
				total_b_transferred += b_written ;
				odp->nextb += b_written ;
			}
			break ;
		}
		/*
		 * everything is ok
		 */
		total_b_transferred += b_transferred ;
		addr += b_transferred ;
	}
	return( total_b_transferred ) ;
}


/*
 * Add a character to a file
 */
int Sputc(int fd, char c)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_od_t *odp = ODP( dp ) ;

#ifdef EVENTS
	EVENT( fd, EV_SPUTC ) ;
#endif

	IO_SETUP( fd, dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;
	ASSERT( odp->start <= odp->nextb && odp->nextb <= odp->buf_end ) ;

	/*
	 * The following is a weak check since we should really be
	 * checking that nextb == buf_end (it would be an error for
	 * nextb to exceed buf_end; btw, the assertion above, when
	 * enabled makes sure this does not occur).
	 *
	 * NOTE: __sio_writef NEVER uses data beyond the end of buffer.
	 */
	if ( odp->nextb >= odp->buf_end )
	{
		int b_in_buffer = odp->buf_end - odp->start ;

		/*
		 * There is nothing we can do if __sio_writef does not manage
		 * to write the whole buffer
		 */
		if ( __sio_writef( odp, fd ) != b_in_buffer )
			return( SIO_ERR ) ;
	}
	*odp->nextb++ = c ;
	if ( __SIO_MUST_FLUSH( *odp, c ) && __sio_writef( odp, fd ) == SIO_ERR )
		return( SIO_ERR ) ;
	return ( c ) ;
}



/*
 * SIO READ FUNCTIONS
 */

/*
 * Stream read call: arguments same as those of read(2)
 */
int Sread(int fd, char *addr, int nbytes)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_id_t *idp = IDP( dp ) ;
	register int b_transferred ;
	int b_read ;
	int total_b_transferred ;
	int b_left ;

#ifdef EVENTS
	EVENT( fd, EV_SREAD ) ;
#endif

	IO_SETUP( fd, dp, __SIO_INPUT_STREAM, SIO_ERR ) ;
	ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;

	b_left = idp->end - idp->nextb ;
	b_transferred = MIN( b_left, nbytes ) ;
	sio_memcopy( idp->nextb, addr, b_transferred ) ;
	idp->nextb += b_transferred ;
	if ( b_transferred == nbytes )
		return( b_transferred ) ;

	nbytes -= b_transferred ;
	total_b_transferred = b_transferred ;
	addr += b_transferred ;

	do
	{
		b_read = __sio_readf( idp, fd ) ;
		switch ( b_read )
		{
			case SIO_ERR:
				if ( total_b_transferred == 0 )
					return( SIO_ERR ) ;
				/* FALL THROUGH */
		
			case 0:
				return( total_b_transferred ) ;
		}
		
		b_transferred = MIN( b_read, nbytes ) ;
		sio_memcopy( idp->nextb, addr, b_transferred ) ;
		addr += b_transferred ;
		idp->nextb += b_transferred ;
		total_b_transferred += b_transferred ;
		nbytes -= b_transferred ;
	}
	while ( nbytes && b_read == idp->buffer_size ) ;
	return( total_b_transferred ) ;
}



/*
 * Read a line from a file
 * Returns a pointer to the beginning of the line or NULL
 */
char *Srdline(int fd)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_id_t *idp = IDP( dp ) ;
	register char *cp ;
	register char *line_start ;
	register int b_left ;
	register int extension ;

#ifdef EVENTS
	EVENT( fd, EV_SRDLINE ) ;
#endif

	IO_SETUP( fd, dp, __SIO_INPUT_STREAM, NULL ) ;
	ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;

#ifdef HAS_MMAP
	if ( idp->memory_mapped && __sio_switch( idp, fd ) == FAILURE )
		return( NULL ) ;
#endif

	b_left = idp->end - idp->nextb ;
	/*
	 * Look for a '\n'. If the search fails, extend the buffer
	 * and search again (the extension is performed by copying the
	 * bytes that were searched to the auxiliary buffer and reading
	 * new input in the main buffer).
	 * If the new input still does not contain a '\n' and there is
	 * more space in the main buffer (this can happen with network
	 * connections), read more input until either the buffer is full
	 * or a '\n' is found.
	 * Finally, set cp to point to the '\n', and line_start to
	 * the beginning of the line
	 */
	if ( b_left && ( cp = sio_memscan( idp->nextb, b_left, '\n' ) ) != NULL )
	{
		line_start = idp->nextb ;
		idp->nextb = cp + 1 ;
	}
	else
	{
		extension = __sio_extend_buffer( idp, fd, b_left ) ;
		if ( extension > 0 )
		{
			ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;

			line_start = idp->start ;
			cp = sio_memscan( idp->nextb, extension, '\n' ) ;
			if ( cp != NULL )
				idp->nextb = cp + 1 ;
			else
				for ( ;; )
				{
					idp->nextb = idp->end ;
					extension = __sio_more( idp, fd ) ;
					if ( extension > 0 )
					{
						cp = sio_memscan( idp->nextb, extension, '\n' ) ;
						if ( cp == NULL )
							continue ;
						idp->nextb = cp + 1 ;
						break ;
					}
					else
					{
						/*
						 * If there is spare room in the buffer avoid trashing
						 * the last character
						 */
						if ( idp->end < &idp->buf[ idp->buffer_size ] )
							cp = idp->end ;
						else
							cp = &idp->buf[ idp->buffer_size - 1 ] ;
						break ;
					}
				}
		}
		else					/* buffer could not be extended */
			if ( b_left == 0 )
			{
				/*
				 * Set errno to 0 if EOF has been reached
				 */
				if ( extension == 0 )
					errno = 0 ;
				return( NULL ) ;
			}
			else
			{
				line_start = idp->start ;
				cp = idp->end ;
				/*
				 * By setting idp->nextb to be equal to idp->end,
				 * subsequent calls to Srdline will return NULL because
				 * __sio_extend_buffer will be invoked and it will return 0.
				 */
				idp->nextb = idp->end ;
			}
	}
	*cp = NUL ;
	idp->line_length = cp - line_start ;
	return( line_start ) ;
}


/*
 * Get a character from a file
 */
int Sgetc(int fd)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_id_t *idp = IDP( dp ) ;

#ifdef EVENTS
	EVENT( fd, EV_SGETC ) ;
#endif

	IO_SETUP( fd, dp, __SIO_INPUT_STREAM, SIO_ERR ) ;
	ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;
	if ( idp->nextb >= idp->end )
	{
		register int b_read = __sio_readf( idp, fd ) ;

		if ( b_read == 0 )
			return( SIO_EOF ) ;
		else if ( b_read == SIO_ERR )
			return( SIO_ERR ) ;
	}
	return( (int) *idp->nextb++ ) ;
}


char *Sfetch(int fd, long int *lenp)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_id_t *idp = IDP( dp ) ;
	register int b_read ;
	register char *p ;

#ifdef EVENTS
	EVENT( fd, EV_SFETCH ) ;
#endif

	IO_SETUP( fd, dp, __SIO_INPUT_STREAM, NULL ) ;
	ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;
	if ( idp->nextb >= idp->end )
	{
		b_read = __sio_readf( idp, fd ) ;
		if ( b_read == SIO_ERR )
			return( NULL ) ;
		if ( b_read == 0 )
		{
			errno = 0 ;
			return( NULL ) ;
		}
	}
	*lenp = idp->end - idp->nextb ;
	p = idp->nextb ;
	idp->nextb = idp->end ;
	return( p ) ;
}



/*
 * SIO CONTROL FUNCTIONS
 */

/*
 * Undo the last Srdline or Sgetc
 */
int Sundo(int fd, int type)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_id_t *idp = IDP( dp ) ;
	int retval = 0 ;

#ifdef EVENTS
	EVENT( fd, EV_SUNDO ) ;
#endif

	CONTROL_SETUP( dp, __SIO_INPUT_STREAM, SIO_ERR ) ;

	/*
	 * Undo works only for fd's used for input
	 */
	if ( dp->stream_type != __SIO_INPUT_STREAM )
		return( SIO_ERR ) ;

	/*
	 * Check if the operation makes sense; if so, do it, otherwise ignore it
	 */
	switch ( type )
	{
		case SIO_UNDO_LINE:
			if ( idp->nextb - idp->line_length > idp->start )
			{
				*--idp->nextb = '\n' ;
				idp->nextb -= idp->line_length ;
			}
			ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;
			break ;
	
		case SIO_UNDO_CHAR:
			if ( idp->nextb > idp->start )
				idp->nextb-- ;
			ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;
			break ;
	
		default:
			retval = SIO_ERR ;
			break ;
	}
	return( retval ) ;
}


/*
 * Flush the buffer associated with the given file descriptor
 * The special value, SIO_FLUSH_ALL flushes all buffers
 *
 * Return value:
 *		0 :  if fd is SIO_FLUSH_ALL or if the flush is successful
 *		SIO_ERR: if fd is not SIO_FLUSH_ALL and
 *		the flush is unsuccessful
 *		or the descriptor is not initialized or it is not
 *		an output descriptor
 */
int Sflush(int fd)
{
   register __sio_descriptor_t *dp ;
   int b_in_buffer ;

#ifdef EVENTS
	EVENT( fd, EV_SFLUSH ) ;
#endif

   if ( fd == SIO_FLUSH_ALL )
   {
      for ( fd = 0, dp = __sio_descriptors ;
				fd < N_SIO_DESCRIPTORS ;
				dp++, fd++ )
         if ( DESCRIPTOR_INITIALIZED( dp ) &&
	      dp->stream_type == __SIO_OUTPUT_STREAM )
            (void) __sio_writef( ODP( dp ), fd ) ;
      return( 0 ) ;
   }
   else
   {
      dp = &__sio_descriptors[ fd ] ;

		CONTROL_SETUP( dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;
      b_in_buffer = ODP( dp )->nextb - ODP( dp )->start ;
      if ( __sio_writef( ODP( dp ), fd ) != b_in_buffer )
         return( SIO_ERR ) ;
      else
         return( 0 ) ;
   }
}


/*
 * Close the file descriptor. This call is provided because
 * a file descriptor may be closed and then reopened. There is
 * no easy way for SIO to identify such a situation, so Sclose
 * must be used.
 *
 * Sclose invokes Sdone which finalizes the buffer.
 * There is no SIO_CLOSE_ALL value for fd because such a thing
 * would imply that the program will exit very soon, therefore
 * the closing of all file descriptors will be done in the kernel
 * (and the finalization will be done by the finalization function
 * NOTE: not true if the OS does not support a finalization function)
 *
 * There is no need to invoke SETUP; Sdone will do it.
 */
int Sclose(int fd)
{
#ifdef EVENTS
	EVENT( fd, EV_SCLOSE ) ;
#endif

	if ( __SIO_FD_INITIALIZED( fd ) )
		if ( Sdone( fd ) == SIO_ERR )
			return( SIO_ERR ) ;
	return( close( fd ) ) ;
}



/*
 * Tie the file descriptor in_fd to the file descriptor out_fd
 * This means that whenever a read(2) is done on in_fd, it is
 * preceded by a write(2) on out_fd.
 * Why this is nice to have:
 * 	1) When used in filters it maximizes concurrency
 *		2) When the program prompts the user for something it forces
 *			the prompt string to be displayed even if it does not
 *			end with a '\n' (which would cause a flush).
 * In this implementation, out_fd cannot be a regular file.
 * This is done to avoid non-block-size write's.
 * The file descriptors are initialized if that has not been done
 * already. If any of them is initialized, it must be for the appropriate
 * stream type (input or output).
 *
 * NOTE: the code handles correctly the case when in_fd == out_fd
 */
int Stie(int in_fd, int out_fd)
{
	struct stat st ;
	register __sio_descriptor_t *dp ;
	int was_initialized ;
	boolean_e failed = NO ;

#ifdef EVENTS
	EVENT( in_fd, EV_STIE ) ;
#endif

	/*
	 * Check if the out_fd is open
	 */
	if ( fstat( out_fd, &st ) == -1 )
		return( SIO_ERR ) ;

	/*
	 * If the out_fd refers to a regular file, the request is silently ignored
	 */
	if ( ( st.st_mode & S_IFMT ) == S_IFREG )
		return( 0 ) ;

	dp = &__sio_descriptors[ in_fd ] ;
	was_initialized = dp->initialized ;		/* remember if it was initialized */
	IO_SETUP( in_fd, dp, __SIO_INPUT_STREAM, SIO_ERR ) ;

	/*
	 * Perform manual initialization of out_fd to avoid leaving in_fd
	 * initialized if the initialization of out_fd fails.
	 * If out_fd is initialized, check if it is used for output.
	 * If it is not initialized, initialize it for output.
	 */
	dp = &__sio_descriptors[ out_fd ] ;
	if ( DESCRIPTOR_INITIALIZED( dp ) )
	{
		if ( dp->stream_type != __SIO_OUTPUT_STREAM )
		{
			failed = YES ;
			errno = EBADF ;
		}
	}
	else
		if ( __sio_init( dp, out_fd, __SIO_OUTPUT_STREAM ) == SIO_ERR )
			failed = YES ;

	if ( failed == NO )
	{
		__SIO_ID( in_fd ).tied_fd = out_fd ;
		return( 0 ) ;
	}
	else
	{
		/*
		 * We failed. If we did any initialization, undo it
		 */
		if ( ! was_initialized )
		{
			int save_errno = errno ;

			(void) Sdone( in_fd ) ;
			errno = save_errno ;
		}
		return( SIO_ERR ) ;
	}
}


/*
 * Untie a file descriptor
 */
int Suntie(int fd)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;

#ifdef EVENTS
	EVENT( fd, EV_SUNTIE ) ;
#endif

	CONTROL_SETUP( dp, __SIO_INPUT_STREAM, SIO_ERR ) ;

	if ( IDP( dp )->tied_fd != SIO_NO_TIED_FD )
	{
		IDP( dp )->tied_fd = SIO_NO_TIED_FD ;
		return( 0 ) ;
	}
	else
	{
		errno = EBADF ;
		return( SIO_ERR ) ;
	}
}


/*
 * Changes the type of buffering on the specified descriptor.
 * As a side-effect, it initializes the descriptor as an output stream.
 */
int Sbuftype(int fd, int type)
{
	register __sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;

#ifdef EVENTS
	EVENT( fd, EV_SBUFTYPE ) ;
#endif

	/*
	 * Check for a valid type
	 */
	if ( type != SIO_LINEBUF && type != SIO_FULLBUF && type != SIO_NOBUF )
	{
		errno = EINVAL ;
		return( SIO_ERR ) ;
	}

	IO_SETUP( fd, dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;
	ODP( dp )->buftype = type ;
	return( 0 ) ;
}


#ifndef sio_memscan

PRIVATE char *sio_memscan( from, how_many, ch )
   char *from ;
   int how_many ;
   register char ch ;
{
   register char *p ;
   register char *last = from + how_many ;

   for ( p = from ; p < last ; p++ )
      if ( *p == ch )
         return( p ) ;
      return( 0 ) ;
}

#endif	/* sio_memscan */


#ifdef NEED_MEMCOPY

void __sio_memcopy( from, to, nbytes )
   register char *from, *to ;
   register int nbytes ;
{
   while ( nbytes-- )
      *to++ = *from++ ;
}

#endif /* NEED_MEMCOPY */


